# Data Scientist Salaries - Data Processing and analysis  

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as skl
from sklearn.preprocessing import LabelEncoder

## Data Preparation

In [3]:
df = pd.read_csv('ds_salaries.csv')
print(df.shape)
print('\n')
df.info()

(3755, 11)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [4]:
df.head()

work_year experience_level employment_type                 job_title   
0       2023               SE              FT  Principal Data Scientist  \
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio   
0   80000             EUR          85847                 ES           100  \
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [5]:
df.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

In [6]:
# Extract categorical variables
cat_vars = ['experience_level', 'employment_type', 'job_title', 'employee_residence', 'company_location', 'company_size']

# Create LabelEncoder object
encoder = LabelEncoder()

# Encode categorical variables as integers
for var in cat_vars:
    df[var] = encoder.fit_transform(df[var])

In [7]:
df.head()

work_year  experience_level  employment_type  job_title  salary   
0       2023                 3                2         84   80000  \
1       2023                 2                0         66   30000   
2       2023                 2                0         66   25500   
3       2023                 3                2         47  175000   
4       2023                 3                2         47  120000   

  salary_currency  salary_in_usd  employee_residence  remote_ratio   
0             EUR          85847                  26           100  \
1             USD          30000                  75           100   
2             USD          25500                  75           100   
3             USD         175000                  11           100   
4             USD         120000                  11           100   

   company_location  company_size  
0                25             0  
1                70             2  
2                70             2  
3                12             1  
4                12             1

In [8]:
df.drop(['salary','salary_currency'] , axis='columns', inplace = True)
df.head()

work_year  experience_level  employment_type  job_title  salary_in_usd   
0       2023                 3                2         84          85847  \
1       2023                 2                0         66          30000   
2       2023                 2                0         66          25500   
3       2023                 3                2         47         175000   
4       2023                 3                2         47         120000   

   employee_residence  remote_ratio  company_location  company_size  
0                  26           100                25             0  
1                  75           100                70             2  
2                  75           100                70             2  
3                  11           100                12             1  
4                  11           100                12             1

## Data Transformation

In [9]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df), columns = df.columns)

## Feature Selection

## Regression - Salary Prediction

In this section, a regression model is going to be trained and tested to predict the salaries of Data Scientists based on the other featuress.

In [10]:
#df_scaled converted to dataframe
df_scaled = pd.DataFrame(df_scaled, columns = df.columns)

# Describing Data Set
df_scaled.describe()

work_year  experience_level  employment_type     job_title   
count  3.755000e+03      3.755000e+03     3.755000e+03  3.755000e+03  \
mean  -1.593130e-13      1.967948e-16     4.475189e-16 -5.109096e-17   
std    1.000133e+00      1.000133e+00     1.000133e+00  1.000133e+00   
min   -3.433303e+00     -2.725008e+00    -1.495172e+01 -2.139921e+00   
25%   -5.404380e-01     -5.178456e-01     2.592668e-02 -6.831569e-01   
50%   -5.404380e-01      5.857357e-01     2.592668e-02 -3.594315e-01   
75%    9.059945e-01      5.857357e-01     2.592668e-02  3.959278e-01   
max    9.059945e-01      5.857357e-01     7.514748e+00  2.823868e+00   

       salary_in_usd  employee_residence  remote_ratio  company_location   
count   3.755000e+03        3.755000e+03  3.755000e+03      3.755000e+03  \
mean    6.055224e-17        2.119329e-16  1.513806e-17     -9.082837e-17   
std     1.000133e+00        1.000133e+00  1.000133e+00      1.000133e+00   
min    -2.100622e+00       -3.439432e+00 -9.524327e-01     -3.550953e+00   
25%    -6.752143e-01        4.601861e-01 -9.524327e-01      4.506246e-01   
50%    -4.076928e-02        4.601861e-01 -9.524327e-01      4.506246e-01   
75%     5.936757e-01        4.601861e-01  1.105918e+00      4.506246e-01   
max     4.955485e+00        5.641760e-01  1.105918e+00      5.077900e-01   

       company_size  
count  3.755000e+03  
mean  -6.812127e-17  
std    1.000133e+00  
min   -2.343022e+00  
25%    2.078761e-01  
50%    2.078761e-01  
75%    2.078761e-01  
max    2.758774e+00

As can be seen, the data is standardly scaled and is ready for regression.

In the following cell, data is split into train and test sets using `sklearn.train_test_split`.

In [34]:
#import train_test_split
from sklearn.model_selection import train_test_split

#seperating features and target
features = df.drop('salary_in_usd', axis=1)
target = df['salary_in_usd'].values

#split dataset into training and test features and targets
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42, shuffle=False)

Scaling the training and test sets in order to start training. The test sets are scaled using the paramaters gained from fitting the scaler to the training sets. This decreases bias in the testing of the model.

In [35]:
from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

#reshaping y_train & y_test to apply the scaling transformation
y_train = y_train.reshape(-1,1)
y_test = y_test.reshape(-1,1)

y_train = scaler.fit_transform(y_train)
y_test = scaler.transform(y_test)


In the following cells, the model is built and tested.

In [36]:
#importing required library to build a neural network model
import tensorflow as tf
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense

In [37]:
#Exploring X_train
print(X_train.shape)
df2 = pd.DataFrame(X_train)
df2.head()

(3004, 8)


0    1    2         3     4    5     6    7
0  0.0  0.0  0.0  2.649351 -49.0  1.0 -45.0 -1.0
1  0.0 -1.0 -2.0  1.714286   0.0  1.0   0.0  1.0
2  0.0 -1.0 -2.0  1.714286   0.0  1.0   0.0  1.0
3  0.0  0.0  0.0  0.727273 -64.0  1.0 -58.0  0.0
4  0.0  0.0  0.0  0.727273 -64.0  1.0 -58.0  0.0

In [38]:
#Exploring y_train
df3 = pd.DataFrame(y_test)
df3.head()

0
0  0.391632
1 -0.065272
2 -0.541679
3 -0.863157
4 -0.300251

A `Sequential` model is built comprising four `Dense` (fully interconnected) layers. It uses the non-linear `RELU` activation function for the hidden layers, and a `linear` output layer. `L2 kernel regularization` is introduced in order to prevent overfitting and promote good generalization to test scenarios.

In [39]:
#constructing model
model = Sequential([
    tf.keras.Input(shape=(8,)),
    Dense(units = 25, activation = 'relu', kernel_regularizer=tf.keras.regularizers.L2(0.001)),
    Dense(units = 10, activation = 'relu', kernel_regularizer=tf.keras.regularizers.L2(0.001)),
    Dense(units = 5, activation = 'relu', kernel_regularizer=tf.keras.regularizers.L2(0.001)),
    Dense(units = 1, activation = 'linear')
])

Model will use the `Adam`(adaptive moment estimation) optimization algorithm, and the `Mean Squared Error` loss function. It is initialized with a `Learning Rate` of 0.0001.

In [44]:
#compiling and summarizing model
from keras.optimizers import Adam
from keras.losses import MeanSquaredError
model.compile(optimizer = Adam(learning_rate = 0.001), loss='mean_squared_error')
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 25)                225       
                                                                 
 dense_5 (Dense)             (None, 10)                260       
                                                                 
 dense_6 (Dense)             (None, 5)                 55        
                                                                 
 dense_7 (Dense)             (None, 1)                 6         
                                                                 
Total params: 546
Trainable params: 546
Non-trainable params: 0
_________________________________________________________________


Now the model is ready to be fit to the data. It is run for a 1000 `epochs`.

In [42]:
history = model.fit(X_train, y_train, validation_split= 0.2 ,epochs = 200)

Epoch 1/200
76/76 [==============================] - 1s 4ms/step - loss: 1.4617 - val_loss: 0.5151
Epoch 2/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5666 - val_loss: 0.4643
Epoch 3/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5513 - val_loss: 0.4633
Epoch 4/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5501 - val_loss: 0.4538
Epoch 5/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5482 - val_loss: 0.4613
Epoch 6/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5371 - val_loss: 0.4594
Epoch 7/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5323 - val_loss: 0.4619
Epoch 8/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5358 - val_loss: 0.4643
Epoch 9/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5338 - val_loss: 0.4598
Epoch 10/200
76/76 [==============================] - 0s 2ms/step - loss: 0.5282 - val_loss: 0.4537
Epoch 11/

Training finished with loss of = 0.4468 on the training set. It can be said that the model has `low bias`, i.e. fits the training data well. However, the validation loss suggests little over fittin, for it has a 3% increase over the training loss. The model will be further tested for generalization and overfitting over the test set.

Forward propogation will be used to predict the target for the test set `y_hat`, which will be compared with the true target `y_test`

In [45]:
y_hat = model.predict(X_test)

24/24 [==============================] - 0s 1ms/step


The `mean_squared_error` function is imported from `sklearn` to be applied between `y_hat` and `y_test`

In [46]:
#
from sklearn.metrics import mean_squared_error
mean_squared_error(y_pred = y_hat, y_true = y_test)

0.5347085573696958

As can be seen, test error is 0.5347, which compared to the training error of 0.0.4468 is a 19.6% increase in error. The model can be said to have high varianve, thus, further improvement is needed.

Real values of `y_hat` and `y_test` can be compared by applying the inverse transformation to the data.

In [47]:
y_hat_real = pd.DataFrame(scaler.inverse_transform(y_hat))
y_test_real = pd.DataFrame(scaler.inverse_transform(y_test))
pd.concat([y_hat_real, y_test_real], axis = 1)

0         0
0    175743.906250  170000.0
1    175743.906250  135000.0
2     95198.796875   98506.0
3     95198.796875   73880.0
4    114473.210938  117000.0
..             ...       ...
746  224238.187500  412000.0
747  126817.992188  151000.0
748   87687.234375  105000.0
749  125564.500000  100000.0
750   92405.843750   94665.0

[751 rows x 2 columns]

It can be seen the model has high error margins. There are few things that can be improved. 
1. The model architecture can be changed, such as the number of hidden layers and/or neurons.
2. The hyperparametres can be adjusted, such the learning rate or regularization parametres.
3. More training epochs might be needed.
4. Further data collection and preperation might be needed.
5. Use a different machine learning algorithm such as XGBoost

### Regression using XGBoost

In this section, an XGBoost learning model will be built to further improve predictions. The `XGBRegressor` class from `xgboost` is used to build the model.

In [48]:
#xgboost library and the XGBRegressor class are imported
import xgboost as xgb
from xgboost import XGBRegressor

`XGBRegressor` is initialized with a squared error loss function for the regression task, and a `random_state` to produce repeatable results.

In [53]:
# define the XGBoost model
xgb_model = XGBRegressor(objective='reg:squarederror', random_state=42)

# train the model using X_train and y_train
xgb_model.fit(X_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=42, ...)

The model is tested on the test set using `xgb_model.predict`, and evaluated using the `sklearn` function `mean_squared_error`.

In [51]:
# make predictions on the test set
y_pred = xgb_model.predict(X_test)

# evaluate the model using mean squared error
mse = mean_squared_error(y_test, y_pred)

print("Mean Squared Error:", mse)

Mean Squared Error: 0.4719166585793818


In [52]:
y_pred_real = pd.DataFrame(scaler.inverse_transform(y_hat))
pd.concat([y_pred_real, y_test_real], axis = 1)

0         0
0    175743.906250  170000.0
1    175743.906250  135000.0
2     95198.796875   98506.0
3     95198.796875   73880.0
4    114473.210938  117000.0
..             ...       ...
746  224238.187500  412000.0
747  126817.992188  151000.0
748   87687.234375  105000.0
749  125564.500000  100000.0
750   92405.843750   94665.0

[751 rows x 2 columns]

As can be seen, results are slightly better suggesting better generalization and less variance(overfitting).